src/votekit/ballot_generator.py

```diff
@@ -977,3 +910,247 @@ def generate_profile(self, number_of_ballots: int) -> PreferenceProfile:
+        blocs = list(self.bloc_voter_prop.keys())
+        bloc_props = list(self.bloc_voter_prop.values())
+        ballots_per_block = dict(
+            zip(blocs, apportion.compute("huntington", bloc_props, number_of_ballots))
+        )
+
+        # dictionary to store preference profiles by bloc
+        pp_by_bloc = {b: PreferenceProfile() for b in blocs}
+
+        for bloc in self.bloc_voter_prop.keys():
+            ballot_pool = []
+            # number of voters in this bloc
+            num_ballots = ballots_per_block[bloc]
+            pref_interval_dict = self.pref_interval_by_bloc[bloc]
+
+            # finds candidates with non-zero preference
+            non_zero_cands = [
+                cand for cand, pref in pref_interval_dict.items() if pref > 0
+            ]
+            # creates the interval of probabilities for candidates supported by this block
+            cand_support_vec = [pref_interval_dict[cand] for cand in non_zero_cands]
+
+            # if there aren't enough non-zero supported candidates, 
+            # include 0 support as ties
+            number_to_sample = self.k
+            number_tied = None
+
+            if len(non_zero_cands) < number_to_sample:
+                number_tied = number_to_sample-len(non_zero_cands)
+                number_to_sample = len(non_zero_cands)
+                
+
+            for _ in range(num_ballots):
+                # generates ranking based on probability distribution of non candidate support
+                # samples k candidates
+                non_zero_ranking = list(
+                    np.random.choice(
+                        non_zero_cands,
+                        number_to_sample,
+                        p=cand_support_vec,
+                        replace=False,
+                    )
+                )
+
+                ranking = [{cand} for cand in non_zero_ranking]
+
+                # if not enough non-zero, add zero support candidates to end as tie
+                zero_cands = list(set(self.candidates).difference(non_zero_cands))
+
+                if number_tied:
+                    tied_candidates = list(
+                        np.random.choice(
+                            zero_cands,
+                            number_tied,
+                            replace=False,
+                        )
+                        )
+                    ranking.append(set(tied_candidates))
+
+                ballot_pool.append(Ballot(ranking=ranking, weight=Fraction(1, 1)))
+            
+            # create PP for this bloc
+            pp = PreferenceProfile(ballots=ballot_pool)
+            pp.condense_ballots()
+            pp_by_bloc[bloc] = pp
+        
+        if by_bloc:
+            return pp_by_bloc
+
+        # else combine the profiles
+        else:
+            pp = PreferenceProfile(ballots=[])
+            for profile in pp_by_bloc.values():
+                pp+= profile
+            return(pp)
```

Comment:
This is an elegant approach but if people want the by_bloc version are they also likely to want the regular profile? And if so would it be better to return both?

Reply:
I think you're right!